### ISSUE TITLE
docker service ps service_name should show container name as well

### ISSUE THREAD

**Issue by sandys** [docker/docker#31369]
_docker ps service_name_ does not show container id or container name. This makes it very unintuitive to figure out the container id of the actual service.

you pretty much have to do a _sudo docker ps_ and try to match the "name" with the name of the service.

What would be very helpful is an additional column that shows container name as well. This will help up run a _docker exec -it container_name /bin/bash_ into it.

**Comment by AkihiroSuda** [comment]
Does `docker ps --filter label=com.docker.swarm.service.name=service_name` work for you?

**Comment by sandys** [comment]
hi Akihiro
We are building tooling around docker including  stats for replication etc
Your command breaks when i have multiple nodes, so i will have to do this
on every node.

Instead if i had this in the service level commands, I know which container
to connect to, on a different node.



On Feb 27, 2017 09:50, "Akihiro Suda" <notifications@github.com> wrote:

Does docker ps --filter label=com.docker.swarm.service.name=service_name
work for you?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/docker/docker/issues/31369#issuecomment-282624667>, or mute
the thread

**Comment by AkihiroSuda** [comment]
`docker exec` is recommended only for debugging purpose. (https://github.com/docker/docker/blob/master/ROADMAP.md#21-docker-exec)

So I think it would be better to "push" the stat data from the service containers to your tool, rather than invoking `docker exec` against the service containers from your tool.

**Comment by sandys** [comment]
hi Akihiro
i accept your comment about docker exec - I want to not focus on that
aspect because it takes away from the real pain point.

Do you see a fundamental issue in "docker service ps" commands giving out
precise container id/name information? This is super super useful in case
of multiple nodes. Otherwise all of us will have to build this tooling by
hand by iterating on each node and running the filter commands.

Its just like the docker system prune commands - everyone had their own
flavor.. whether or not that was correct.

On Feb 27, 2017 10:08, "Akihiro Suda" <notifications@github.com> wrote:

docker exec is recommended only for debugging purpose. (
https://github.com/docker/docker/blob/master/ROADMAP.md#21-docker-exec)

So I think it would be better to "push" the stat data from the service
containers to your tool, rather than invoking docker exec against the
service containers from your tool.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/docker/docker/issues/31369#issuecomment-282626321>, or mute
the thread

**Comment by sandys** [comment]
Also just to clarify in case you were concerned about simplicity of
information... you can hide this behind an optional switch "docker service
ps name -a" or something.

Anything you feel is more appropriate as long as they are available at the
service level and in one place (regardless of all the nodes)

**Comment by AkihiroSuda** [comment]
If you really need this feature, `for f in $(docker service ps -q $service);do docker inspect --format '{{.NodeID}} {{.Status.ContainerStatus.ContainerID}}' $f; done` would work for you.

> Do you see a fundamental issue in "docker service ps" commands giving out precise container id/name information? 

It would mislead people to use `docker exec` for non-debugging purpose.

**Comment by sandys** [comment]
your script shows all containers for a service - **including those that have shutdown**

With all due respect, I hope it illustrates the non-trivial nature of this task. People like me only want to be empowered with the correct information without prejudice about second guessing how I am going to potentially use this. It is my grave to dig... atleast show me the right spot to dig ;)

> It would mislead people to use docker exec for non-debugging purpose.

**Comment by thaJeztah** [comment]
I guess we can add container properties to the "task" context, so that it can be used as part of the `--format` template? https://github.com/docker/docker/blob/f8c7c921d905615acf1d2c6577ae1cfb4600bec0/cli/command/formatter/task.go#L15-L145

@AkihiroSuda wdyt?

**Comment by sandys** [comment]
i also want to mention that would mean *all* containers related to a particular service - even if they are running on different machines. because the master already has this information. 

i think the earlier script will only show the containers running on the machine that the script is run on(i dont have a multinode cluster to play with right now to verify)

**Comment by thaJeztah** [comment]
Yes, `docker service ps` shows all tasks for a service, so adding the container name/id to the list of options that can be used in `--format` would address that use case

**Comment by AkihiroSuda** [comment]
@sandys

> your script shows all containers for a service - including those that have shutdown
You should be able to use `--filter`. Also, it works with the containers running on other nodes as well.

@thaJeztah 

> I guess we can add container properties to the "task" context, so that it can be used as part of the --format template? 

If we really need this feature, I'm ok to add this feature (probably it should be some annotation).
But I'm not sure (yet) this feature is needed.

**Comment by sandys** [comment]
@sanimej thanks for this info! so what is the right way to do this ? "docker service" command or "docker network inspect" command.
Although this is a great feature, im kind of wondering if the service command is the right place for this ?

**Comment by thaJeztah** [comment]
The feature in docker 17.04 will be on `docker network inspect`

**Comment by sanimej** [comment]
@sandys https://github.com/docker/docker/pull/31710#issuecomment-285553114 explains the rationale.

**Comment by sandys** [comment]
I just got the docker update where this fix is present - is there a way to
format the output so that I have the container ids of all replicas of a
service ? I'm not able to figure out how to do that.

**Comment by koubas** [comment]
One way is to get the task name and id and concatenate them using "." as a separator: `docker service ps <service> --no-trunc --format "{{.Name}}.{{.ID}} "`, but this might be considered misusing the internal implementation of how the container names are generated.

**Comment by freakabcd** [comment]
I had to do exactly this recently. To summarise:
`service=myservicename; for f in $(docker service ps -q --filter desired-state=running $service); do docker inspect --format '{{.NodeID}} {{.Status.ContainerStatus.ContainerID}}' $f; done`
works as expected to provide running containerIDs (and associated nodeIDs) for a given service name.

I don't mind if the required information is not available directly from a `docker service ...` command.
Getting the information from `docker network inspect --verbose` is a bit weird as I would expect to use the output for debugging network issues rather than figuring out containers for a particular service.

**Comment by thaJeztah** [comment]
Container names are composed of `Name` and `ID` (non-truncated), so it's possible to show the name of the container with a custom format;


```
docker service ps myservice --format 'table {{.ID}}\t{{.Name}}\t{{.Name}}.{{.ID}}\t{{.Image}}\t{{.Node}}\t{{.DesiredState}}\t{{.CurrentState}}' --no-trunc
ID                          NAME                NAME.ID                                     IMAGE                                                                                  NODE                    DESIRED STATE       CURRENT STATE
8b4p2h7ows9uhuhqfjik79i09   myservice.1         myservice.1.8b4p2h7ows9uhuhqfjik79i09       nginx:alpine@sha256:3a44395131c5a9704417d19ab4c8d6cb104013659f5babb2f1c632e789588196   linuxkit-025000000001   Running             Running 3 minutes ago
ro0huws43gxecgtudjvgxr1v8    \_ myservice.1      \_ myservice.1.ro0huws43gxecgtudjvgxr1v8   nginx:alpine@sha256:3a44395131c5a9704417d19ab4c8d6cb104013659f5babb2f1c632e789588196   linuxkit-025000000001   Shutdown            Failed 3 minutes ago
uhibvsksi9v4ex9qax4uxickn    \_ myservice.1      \_ myservice.1.uhibvsksi9v4ex9qax4uxickn   nginx:alpine@sha256:3a44395131c5a9704417d19ab4c8d6cb104013659f5babb2f1c632e789588196   linuxkit-025000000001   Shutdown            Shutdown 3 minutes ago
qo83kkfduotfsg2klx4fjpt4s    \_ myservice.1      \_ myservice.1.qo83kkfduotfsg2klx4fjpt4s   nginx:alpine@sha256:3a44395131c5a9704417d19ab4c8d6cb104013659f5babb2f1c632e789588196   linuxkit-025000000001   Shutdown            Shutdown 3 weeks ago
yyh9r66ckz7r54uzjc82vvzti    \_ myservice.1      \_ myservice.1.yyh9r66ckz7r54uzjc82vvzti   nginx:alpine@sha256:3a44395131c5a9704417d19ab4c8d6cb104013659f5babb2f1c632e789588196   linuxkit-025000000001   Shutdown            Shutdown 4 weeks ago
```

I don't think the CLI configuration file currently supports setting the custom format for `service ps`, but this could likely be added; https://docs.docker.com/engine/reference/commandline/cli/#configuration-files

**Comment by freakabcd** [comment]
@thaJeztah ah its true that one can get to the container using its name. @koubas had mentioned the exact same thing earlier, I completely missed it as I was attempting to find the container IDs.
